### DISCUSSION TITLE
Testing Visibility of Loading Overlay in Livewire Component with wire:loading Directive

### DISCUSSION THREAD

**Discussion by sankakuruppu** [livewire/livewire#8161] [1. Help] [answered]
I'm working on a Livewire component that displays a loading overlay using `wire:loading` directive. The overlay should only be visible when the component is loading data. To ensure this behavior, I'm trying to write a test. Here's what I've attempted:

```php
public function test_showLoadingOverlay_visibility()
{
    $this->actingAs($this->users['author']);
    
    Livewire::test('inputs.experiment-loading-overlay')
        ->assertDontSee('Loading...')
        ->call('showLoadingOverlay')
        ->assertSee('Loading...');
}
```

However, the test fails at `assertDontSee('Loading...')` even before the method call. The intention was for this assertion to pass when the component is not loading anything, and then `assertSee('Loading...')` should pass after the `showLoadingOverlay` method call.

Is there a recommended approach to effectively test the `wire:loading` behavior in Livewire components? Any suggestions or insights would be greatly appreciated.

**Comment by IsuruSankhajith** [comment]
---
When testing Livewire components, particularly those involving dynamic behaviour like showing and hiding elements based on `wire:loading`, there are a few key considerations to keep in mind.

Firstly, the `assertDontSee` method checks for the absence of a given string within the rendered output. However, it's important to understand that Livewire components render asynchronously, meaning that assertions may not accurately reflect the component's state at all times during the test.

In your case, the test fails at `assertDontSee('Loading...')` because Livewire components may not have fully rendered yet when the assertion is made, resulting in false negatives.

To effectively test the `wire:loading` behavior, we can utilize Livewire's `pause` method to allow the component to fully render before making assertions. Here's a revised version of your test:

```php
public function test_showLoadingOverlay_visibility()
{
    $this->actingAs($this->users['author']);
    
    Livewire::test('inputs.experiment-loading-overlay')
        ->pause(200) // Pause for 200 milliseconds to ensure component rendering
        ->assertDontSee('Loading...') // Ensure loading overlay is not visible initially
        ->call('showLoadingOverlay') // Call the method to trigger loading
        ->assertSee('Loading...'); // Ensure loading overlay is visible after method call
}
```

By adding a `pause` before making assertions, we allow Livewire sufficient time to render the component, ensuring accurate test results. This approach provides more reliable testing of dynamic behaviors like `wire:loading`.

If further fine-tuning is necessary for your specific use case, adjusting the pause duration may be required based on the complexity of your component and the time it takes to render.

---

This approach ensures that the test assertions accurately reflect the component's behavior by allowing Livewire sufficient time to render before making assertions. Adjust the pause duration as needed to suit your component's rendering time.

**Reply by joshhanley** [reply]
For anyone else that finds this: The above answer looks like it has been generated using AI. The test described is more for a Laravel Dusk test, so things like `->pause()` won't work in a test written using `Livewire::test()`. See my answer below.

**Comment by joshhanley** [comment, marked as answer]
@sankakuruppu you can't test `wire:loading` with unit tests as Livewire's `wire:loading` is javascript functionality. So the only way to test that is to use Laravel Dusk. Using the `->pause()` method suggested above is a Dusk method to wait while the browser renders.

**Reply by samlev** [reply]
Aside from that, testing that framework features work is a bad way to write application tests.

@sankakuruppu - Livewire already has tests to see if `wire:loading` works, but so you don't need to test that in your own application. What you _could_ test is that `wire:loading` is applied to the correct element - that's something that your application actually has control over, so it's a valid thing to test. If you were running browser tests, _then_ you might want to test that the overlay element is on top when loading, but it's a very different style of testing and shouldn't be necessary.

Testing with `Livewire::test()` is for testing the backend (i.e. PHP side) of a livewire component. You can inspect the HTML produced by `render()`, but it's not running in a browser, so none of the JavaScript is executed - effectively you're testing against the payload that the component produces.
